File: Data Structures & Algorithms/quickSort/submission-2.py
# Definition for a pair.
# class Pair:
#     def __init__(self, key: int, value: str):
#         self.key = key
#         self.value = value
class Solution:
    def quickSort(self, pairs: List[Pair]) -> List[Pair]:
        if len(pairs) <= 1:
            return pairs
        
        def quickSortHelper(s, e):
            if e - s + 1 <= 1:
                return
            pivot = pairs[e].key
            left = s
            for i in range(s, e):
                if pairs[i].key < pivot:
                    tmp = pairs[i]
                    pairs[i] = pairs[left]
                    pairs[left] = tmp
                    left += 1
            # swap left and pivot
            tmp = pairs[left]
            pairs[left] = pairs[e]
            pairs[e] = tmp
            # recursive call
            quickSortHelper(s, left - 1)
            quickSortHelper(left + 1, e)

        quickSortHelper(0, len(pairs) - 1)
        return pairs

File: Data Structures & Algorithms/insertionSort/submission-2.java
// Definition for a pair
// class Pair {
//     int key;
//     String value;
//
//     Pair(int key, String value) {
//         this.key = key;
//         this.value = value;
//     }
// }
public class Solution {
    public List<List<Pair>> insertionSort(List<Pair> pairs) {
        List<List<Pair>> res = new ArrayList<>();

        for (int i = 0; i < pairs.size(); i++) {
            int j = i - 1;
            while (j >= 0 && pairs.get(j + 1).key < pairs.get(j).key) {
                swap(j, j + 1, pairs);
                j -= 1;
            }
            List<Pair> clone = new ArrayList<>(pairs);
            res.add(clone);
        }
        return res;
    }

    public void swap(int x, int y, List<Pair> list) {
        Pair temp = list.get(x);
        list.set(x, list.get(y));
        list.set(y, temp);
    }
}
